book/content/reproducible_environments/09/case-study-branchLSTM.md

```diff
@@ -0,0 +1,110 @@
+In the third, the model with the optimal hyperparameters is applied to the test dataset to obtain predictions of whether each tweet is supporting, denying, querying or commenting on the rumour in the original root tweet.
+Elena had saved the values of the hyperparameters used for her final run of the code, which was incredibly helpful for checking reproducibility as it was easy to isolate the two stages that involved the branchLSTM model itself.
+
+My first task was to run the final stage of the code on resources other than the cluster that Elena had previously used.
+I ran my first tests on the computers I had available (an iMac and MacBook Pro) and an NC6 virtual machine from Microsoft Azure, all of which yielded different overall accuracy values to those in Elena's paper.
+
+At this stage, such a result wasn't unexpected - I had used the most up-to-date versions of the Python packages used in the project, whereas most of Elena's work on the project had taken place over a year earlier.
+But the results were also different between the three computers, so before getting stuck into how the package versions affected the model, I decided to check the earlier stages of the script, where the data preprocessing took place.
+
+This uncovered some of the most interesting differences in interoperability that I've ever come across!
+
+## My computer is not the same as your computer
+
+Each tree of tweets was stored in a separate file, within a folder concerning the theme of those tweets.
+The lists of these files were identified by using Python's `os.listdir()` command to list all the files within the relevant data directories.
+The order of the output of `listdir()` is actually arbitrary, but this isn’t apparent if you always work on the same machine, as Elena had done.
+While the same data overall was being read in on different computers, it was concatenated together in a different order and so the data provided to the model was not exactly the same.
+This led to slight differences in model accuracy.
+
+Sorting the list after generating it solved this problem, and it fortunately seemed that the default sort order on the cluster used by Elena had been alphabetical (as it is on many systems), meaning that we could generate the same order used originally.
+
+## How long is a smiley face?
+
+Once loaded, numerical representations of each tweet were generated as 314-element arrays.
+The first 14 elements stored quantifiable features of the tweet, such as the number of negation words and the presence of attachments, and the remaining 300 elements stored a [Word2vec](https://skymind.ai/wiki/word2vec) representation of the tweet's textual content.
+When I compared these features across machines, I noticed something rather odd - the contents of corresponding elements were identical, apart from those which stored the length (to be more precise, the character count) of some of the tweets.
```

Comment:
Oh yeah! And here again - was this a manual thing, where you transferred CSVs to one computer? Was there a comparison tool or did you write some code?

Reply:
This was manual, but without much code - the only thing I added in the codebase itself was a temporary print statement to show the vector representations on the terminal. I piped that output to a file, then used `diff` for the comparison.

I've added a little note saying that I printed the arrays and compared them, do you think that's enough?

Reply:
Yeah! That's great :)